index.html

```diff
@@ -3737,55 +3739,53 @@ <h2>DID Resolution</h2>
+A <a href="#metadata-structure">metadata structure</a> containing properties
+defined in <a href="#did-resolution-options"></a>. This input is
+REQUIRED, but the structure MAY be empty.
             </dd>
         </dl>
 
         <p>
-The output variables of these functions are as follows:
+The return value of <code>resolve</code> is the
+<a>didResolutionMetadata</a>, <a>didDocument</a>, and
+<a>didDocumentMetadata</a> properties. The return value of
+<code>resolveRepresentation</code> is the
+<a>didResolutionMetadata</a>, <a>didDocumentStream</a>, and
+<a>didDocumentMetadata</a> properties. These properties are described below:
         </p>
 
         <dl>
             <dt>
-did-resolution-metadata
+<dfn>didResolutionMetadata</dfn>
             </dt>
             <dd>
-              <p>
 A <a href="#metadata-structure">metadata structure</a> consisting of values
-relating to the results of the <a>DID resolution</a> process. This structure
-is REQUIRED and MUST NOT be empty.
-              </p>
-              <p>
-This metadata typically changes between invocations of the
-<code>resolve</code> and <code>resolveRepresentation</code> functions as it
-represents data about the resolution process itself. Properties defined by
-this specification are in <a href="#did-resolution-metadata-properties"></a>.
-              </p>
-              <p>
-If the resolution is successful, and if the <code>resolveRepresentation</code>
-function was called, this structure MUST contain a <code>contentType</code>
-property containing the mime-type of the <code>did-document-stream</code> in
-this result. If the resolution is not successful, this structure MUST contain
-an <code>error</code> property describing the error.
-              </p>
+relating to the results of the <a>DID resolution</a> process which typically
+changes between invocations of the <code>resolve</code> and
+<code>resolveRepresentation</code> functions, as it represents data about the
+resolution process itself. This structure is REQUIRED, and MUST NOT be empty.
+This metadata is defined by <a href="#did-resolution-metadata"></a>. If
+<code>resolveRepresentation</code> was called, this structure MUST contain a
+<code>contentType</code> property containing the MIME type of the
+<code>didDocument</code>. If the resolution is not successful, this structure
+MUST contain an <code>error</code> property describing the error.
             </dd>
             <dt>
-did-document
+<dfn>didDocument</dfn>
             </dt>
             <dd>
 If the resolution is successful, and if the <code>resolve</code> function was
-called, this MUST be a <a>conforming DID document</a>. If the resolution is
+called, this MUST be a <a>DID document</a> (abstract data model) as described in
+<a href="#data-model"></a> that is capable of being transformed
+into a <a>conforming DID Document</a> (representation), using the production
+rules specified by the representation. If the resolution is
```

Comment:
I don't understand why this is requirement was added here. If I can get it back into the ADM, then the resolve is done and whether I can serialize it back out is irrelevant to this function.

```suggestion
<a href="#data-model"></a>. If the resolution is
```

Reply:
@jricher my understanding is that it is useful to say this here in order to make the resolve() function testable (see the title of the PR). I agree with @msporny 's proposed addition here.

Reply:
The language that was here before didn't say you go back to the ADM... it said that you go back to a representation. The reference to "conforming DID Document" was confusing, because conforming DID Documents are expected to be expressed in a representation... which was not the intent here. So, the change was made to make sure we were talking about a `didDocument` in the abstract sense... but made sure that was testable by saying you had to be able to take whatever that was and get it out to a representation (in order to make it clearly testable).

Reply:
I do not agree that this makes it any more "testable" than before. The burden is on the test suite to serialize the result in either case, and a conformant DID document already needs to be representable. This requirement is redundant and implies that there are DID documents that could possibly :not: be representable in some form.

Reply:
The change makes it explicit that the return value is in ADM form, and that it must be serializable in a representation. By doing so, we achieve what you want (which is for the return value to be abstract), but also enable the return value to be testable by the test suite (and therefore usable by an implementation).

@jricher, if you disagree with this change, we need a counter-proposal from you on how we get this particular return value to testability, or we need you to clarify if you are objecting or if you can live with the language.

Reply:
An alternative proposal to testability is to fix the problem in the test suite itself:

To connect to the test suite, and implementation has to define a concrete means for the test suite to call the `resolve`, `resolveResolution`, and `dereference` functions, including concrete means for supplying all inputs and receiving all outputs. 

This is text that does not belong in the specification proper (in my opinion) but makes it clear what harnesses the test suite will need to test implementations of these functions.

Reply:
Another way to look at it is that this "testability" push is turning this functional definition section into a _protocol_. It is my belief that this is explicitly not a protocol, and as such calling these functions in an implementation might require a network call, writing files to a directory, linking to a library and calling a native function, running a utility script, or any number of possible ways to call a function. This is the problem that I have had with putting the burden on the test suite to test implementations of these functions in this way. If we were defining a protocol, the arguments around serializability and normativity would make sense, but as it is they do not and I believe this is a result of overreach from the test suite, which needs to define its own requirements for common testability, effectively creating a protocol on top of these functions for the purposes of the test suite.

Reply:
> To connect to the test suite, and implementation has to define a concrete means for the test suite to call the resolve, resolveResolution, and dereference functions, including concrete means for supplying all inputs and receiving all outputs.

This is typically done by W3C Working Groups by defining the concrete API in the specification and writing a test suite against that concrete API. Implementers then expose that concrete API in their implementations:

https://w3c.github.io/FileAPI/#file-section

We are being far more lax with the test-ability of the Resolution section by doing a hand-wave over the API and saying that a serialization will be available to the test suite to test.

That said, your opinion is noted and there is a record of it.